src/sqlfluff/rules/std.py

```diff
@@ -2827,3 +2827,92 @@ def _eval(self, segment, **kwargs):
                         description=("Found USING statement. Expected only ON statements.")
                     )]
         return None
+
+
+@std_rule_set.register
+class Rule_L034(BaseCrawler):
+    """Use wildcards then simple select targets before calculations and aggregates.
+
+    | **Anti-pattern**
+
+    .. code-block:: sql
+
+        select
+            a,
+            *,
+            row_number() over (partition by id order by date) as y,
+            b
+        from x
+
+
+    | **Best practice**
+    |  Order "select" targets in ascending complexity
+
+    .. code-block:: sql
+
+        select
+            *,
+            a,
+            b,
+            row_number() over (partition by id order by date) as y
+        from x
+
+    """
+
+    def _validate(self, i, segment):
+        # Check if we've seen a more complex select target element already
+        if self.seen_element_band[i + 1::] != [None] * len(self.seen_element_band[i + 1::]):
+            # Not quite worked out the fix
+            # earliest_bad_element = next(el for el in self.seen_element_band[i+1::] if el is not None)
+            self.violation_buff.append(
+                LintResult(
+                    anchor=segment,
+                    # fixes=[LintFix('create', segment, earliest_bad_element),
```

Comment:
@alanmcruickshank I'm trying to 'swap' the elements here to fix the order. This commented out lintfix results in:

```
select
    a,
    *,
    row_number() over (partition by id order by date) as y,
    b
from x
```

going to

```
select
    *a,
    ,
    brow_number() over (partition by id order by date) as y,
    
from x
```


any ideas on how to solve?

Reply:
I think I need to somehow consume all the new segments up to and including the comma, but then it gets a bit complex with the final select target statement not having a comma 🤔

Reply:
Suggestion:

We really only have to identify three "chunks".
1. A series of segments which contain the aggregate elements and the commas between them.
2. The comma separating the aggregates and the fields,
3. A series of segments which contain the fields after that comma and the commas between them.

Any non code elements which are before 1 or after 3 are excluded. Any non code elements between 1 & 2 or between 2 & 3 should belong to 2.

Once we've identified those chunks, we just want to swap them. That should at the very least _preserve_ the layout (although it won't fix the layout if it was already bad - but I don't think that's what this rule is for).

To do that swap we have two options.
1. We _could_ implement a type of `LintFix` called `swap`, to do exactly this. I think that could be a useful feature but I think we'd need to do a bit of thinking around how to do it well.
2. We instead return a series of fixes: 2 `edit` fixes, where we anchor on the first segment of chunk 1, and edit it to all the segments of chunk 3 (and visa versa, anchored on the first segment of chunk 3 with all the segments of chunk 1). + a series of `delete` fixes to remove all of the elements in chunk 1 or 3 that aren't the first one (which is targeted for the edit).

Sounds a little complicated, but I suspect that's probably the best way of doing it.

Any ideas?

Reply:
Thanks for the suggestions. Thinking about it more - it seems like really all we need to do is straight swaps. What's the reason that the swap wouldn't work if I just produce two edits like I'm doing now? This works for the example at the start of the thread, but not in the second below.

```
fixes=[LintFix('edit', earliest_bad_element, segment),
           LintFix('edit', segment, earliest_bad_element)]
)
```

This example now works:

```
select
    a,
    *,
    row_number() over (partition by id order by date) as y,
    b
from x
```

goes to

```
select
    *,
    a,
    b,
    row_number() over (partition by id order by date) as y
from x
```

but this one doesn't:

```
select
    row_number() over (partition by id order by date) as y,
    *,
    cast(b as int) as b_int
from x
```

is going to

```
select
    *,
    row_number() over (partition by id order by date) as y,
    row_number() over (partition by id order by date) as y
from x
```

Reply:
I *think* the issue is that the lint fix isn't designed to reference *multiple* segments in the first argument of the edit. However on reflection, that's not an awful idea for cases like this. It would however probably mean a bit of a refactor to deal with that eventuality.

The `anchor` attribute is used in the matching process to work out where to apply the fix, and if in your case it's now a tuple of segments rather than an individual, then it might behave strangely.

Worth confirming for yourself that that is what's happening though.